tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,520 @@
+            return Ok((
+                leader_tpu_map.take().unwrap(),
+                epoch_info.take().unwrap(),
+                slot_leaders.take().unwrap(),
+            ));
+        }
+
+        num_attempts = num_attempts.saturating_add(1);
+        sleep(ATTEMPTS_SLEEP_DURATION).await;
+    }
+    Err(Error::InitializationFailed)
+}
+
+fn adjust_lookahead(estimated_slot: Slot, slot_leaders: &SlotLeaders, lookahead_leaders: u8) -> u8 {
+    if slot_leaders
+        .is_last_slot_in_window(estimated_slot)
+        .unwrap_or(false)
+    {
+        lookahead_leaders.saturating_add(1)
+    } else {
+        lookahead_leaders
+    }
+}
+
+async fn try_update<F, Fut, T>(
+    label: &str,
+    data: &mut T,
+    make_call: F,
+    num_failures: &mut usize,
+    max_failures: usize,
+) -> Result<(), Error>
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = Result<T, Error>>,
+{
+    match make_call().await {
+        Ok(result) => {
+            *num_failures = 0;
+            debug!("{label} updated successfully");
+            *data = result;
+            Ok(())
+        }
+        Err(e) => {
+            *num_failures = num_failures.saturating_add(1);
+            warn!("Failed to update {label}: {e} ({num_failures} consecutive failures)",);
+
+            if *num_failures >= max_failures {
+                error!("Max consecutive failures for {label}, giving up.");
+                Err(e)
+            } else {
+                Ok(())
+            }
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct LeaderTpuMap {
+    last_cluster_refresh: Instant,
+    leader_tpu_map: HashMap<Pubkey, SocketAddr>,
+}
+
+impl LeaderTpuMap {
+    async fn new(rpc_client: &impl ClusterInfoProvider) -> Result<Self, Error> {
+        let leader_tpu_map = rpc_client.leader_tpu_map().await?;
+        Ok(Self {
+            last_cluster_refresh: Instant::now(),
+            leader_tpu_map,
+        })
+    }
+
+    fn get(&self, leader: &Pubkey) -> Option<&SocketAddr> {
+        self.leader_tpu_map.get(leader)
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct SlotLeaders {
+    first_slot: Slot,
+    leaders: Vec<Pubkey>,
+}
+
+impl SlotLeaders {
+    async fn new(
+        rpc_client: &impl ClusterInfoProvider,
+        estimated_current_slot: Slot,
+        slots_in_epoch: Slot,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            first_slot: estimated_current_slot,
+            leaders: rpc_client
+                .slot_leaders(estimated_current_slot, slots_in_epoch)
+                .await?,
+        })
+    }
+
+    fn last_slot(&self) -> Slot {
+        self.first_slot + self.leaders.len().saturating_sub(1) as u64
+    }
+
+    fn slot_leader(&self, slot: Slot) -> Option<&Pubkey> {
+        slot.checked_sub(self.first_slot)
+            .and_then(|index| self.leaders.get(index as usize))
+    }
+
+    fn is_last_slot_in_window(&self, slot: Slot) -> Option<bool> {
```

Comment:
nit: this will fail for the very last slot, which technically **is** _is_last_slot_in_window_, but method will return `None` in that case which is misleading/wrong